pkg/cloud/rgraph/exec/executor_parallel.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2024 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exec
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud"
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/rgraph/algo"
+	"k8s.io/klog/v2"
+)
+
+var (
+	ErrPendingActions = errors.New("Executor did not process all actions")
+)
+
+func defaultParallelExecutorConfig() *ExecutorConfig {
+	return &ExecutorConfig{
+		DryRun:        false,
+		ErrorStrategy: ContinueOnError,
+		Timeout:       5 * time.Minute,
+	}
+}
+
+// NewParallelExecutor returns a new Executor that runs tasks multi`-threaded.
+func NewParallelExecutor(pending []Action, opts ...Option) (*parallelExecutor, error) {
+	ret := &parallelExecutor{
+		config: defaultParallelExecutorConfig(),
+		result: &Result{Pending: pending},
+		pq:     algo.NewParallelQueue[Action](),
+	}
+	for _, opt := range opts {
+		opt(ret.config)
+	}
+
+	if err := ret.config.validate(); err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+type parallelExecutor struct {
+	config *ExecutorConfig
+	c      cloud.Cloud
+
+	// lock guards results
+	lock   sync.Mutex
+	result *Result
+
+	pq   *algo.ParallelQueue[Action]
+	done chan *TraceEntry
+}
+
+// parallelExecutor implements Executor.
+var _ Executor = (*parallelExecutor)(nil)
+
+func (ex *parallelExecutor) Run(ctx context.Context, c cloud.Cloud) (*Result, error) {
+	ex.queueRunnableActions()
+	subctx, cancel := context.WithTimeout(ctx, ex.config.Timeout)
+
+	runAction := func(ctx context.Context, a Action) error {
+		te := &TraceEntry{
+			Action: a,
+			Start:  time.Now(),
+		}
+		klog.V(4).Infof("Run action %s", a)
+		events, runErr := a.Run(ctx, c)
+		te.End = time.Now()
+		klog.V(4).Infof("Finish action %s, err: %v", a, runErr)
+
+		ex.addActionResult(a, runErr)
+
+		if runErr != nil {
+			klog.Infof("Got error  %v, from action %s error_strategy: %s", runErr, a, ex.config.ErrorStrategy)
+			// check error strategy and decide if new actions should be executed.
+			if ex.config.ErrorStrategy == StopOnError {
+				return fmt.Errorf("parallelExecutor: stopping execution for Action %s (got %w)", a, runErr)
```

Comment:
* We need to be careful -- returning here measn the Tracer does not record the error.
* suggest the error be more direct: 

```
return fmt.Errorf("parallelExecutor: StopOnError due to Action %s: %w", a, runErr)
```

Reply:
done